From b5bde3c2182322ce8c99aa8a53c069cc094a7093 Mon Sep 17 00:00:00 2001
From: nobody <nobody@github.com>
Date: Wed, 20 Jan 2021 19:57:24 +0800
Subject: [PATCH] rockchip: Add Nanopi R4S board support

Add initial support for Nanopi R4S board.

Specification
- Rockchip RK3399
- LPDDR4 4GiB, 1GB DDR3-1866
- SD card slot
- RTL8211E 1Gbps
- RTL8111H 1Gbps
- USB 3.0 x2
- USB Type C power

Signed-off-by: jensen <jensenhuang@friendlyarm.com>
---
 target/linux/rockchip/Makefile                |   3 +-
 target/linux/rockchip/image/armv8.mk          |  10 +
 .../rockchip/image/nanopi-r4s.bootscript      |   8 +
 ...-initial-signal-voltage-on-power-off.patch |  32 +++
 ...ew-binding-to-specify-host-min-index.patch |  38 +++
 ...ockchip-Add-Nanopi-R4S-board-support.patch | 239 ++++++++++++++++++
 6 files changed, 329 insertions(+), 1 deletion(-)
 create mode 100644 target/linux/rockchip/image/nanopi-r4s.bootscript
 create mode 100644 target/linux/rockchip/patches-5.4/300-mmc-core-set-initial-signal-voltage-on-power-off.patch
 create mode 100644 target/linux/rockchip/patches-5.4/301-mmc-add-new-binding-to-specify-host-min-index.patch
 create mode 100644 target/linux/rockchip/patches-5.4/302-arm64-dts-rockchip-Add-Nanopi-R4S-board-support.patch

diff --git a/target/linux/rockchip/Makefile b/target/linux/rockchip/Makefile
index d3608f5fb..35a25da9d 100644
--- a/target/linux/rockchip/Makefile
+++ b/target/linux/rockchip/Makefile
@@ -6,8 +6,9 @@ include $(TOPDIR)/rules.mk
 
 BOARD:=rockchip
 BOARDNAME:=Rockchip
-FEATURES:=ext4 audio usb usbgadget display gpio fpu rootfs-part boot-part squashfs
+FEATURES:=ext4 audio usb usbgadget display gpio fpu pci pcie rootfs-part boot-part squashfs
 SUBTARGETS:=armv8
+CPU_TYPE:=cortex-a53
 
 KERNEL_PATCHVER=5.4
 
diff --git a/target/linux/rockchip/image/armv8.mk b/target/linux/rockchip/image/armv8.mk
index 8e4ba07e0..f8b6a0c92 100644
--- a/target/linux/rockchip/image/armv8.mk
+++ b/target/linux/rockchip/image/armv8.mk
@@ -23,3 +23,13 @@ define Device/pine64_rockpro64
   IMAGE/sysupgrade.img.gz := boot-common | boot-script | pine64-img | gzip | append-metadata
 endef
 TARGET_DEVICES += pine64_rockpro64
+
+define Device/friendlyarm_nanopi-r4s
+  DEVICE_VENDOR := FriendlyARM
+  DEVICE_MODEL := NanoPi R4S
+  SOC := rk3399
+  UBOOT_DEVICE_NAME := nanopi-r4s-rk3399
+  IMAGE/sysupgrade.img.gz := boot-common | boot-script nanopi-r4s | pine64-img | gzip | append-metadata
+  DEVICE_PACKAGES := kmod-usb-net-rtl8152 kmod-r8168
+endef
+TARGET_DEVICES += friendlyarm_nanopi-r4s
diff --git a/target/linux/rockchip/image/nanopi-r4s.bootscript b/target/linux/rockchip/image/nanopi-r4s.bootscript
new file mode 100644
index 000000000..abe9c24ee
--- /dev/null
+++ b/target/linux/rockchip/image/nanopi-r4s.bootscript
@@ -0,0 +1,8 @@
+part uuid mmc ${devnum}:2 uuid
+
+setenv bootargs "console=ttyS2,1500000 earlycon=uart8250,mmio32,0xff1a0000 root=PARTUUID=${uuid} rw rootwait"
+
+load mmc ${devnum}:1 ${fdt_addr_r} rockchip.dtb
+load mmc ${devnum}:1 ${kernel_addr_r} kernel.img
+
+booti ${kernel_addr_r} - ${fdt_addr_r}
diff --git a/target/linux/rockchip/patches-5.4/300-mmc-core-set-initial-signal-voltage-on-power-off.patch b/target/linux/rockchip/patches-5.4/300-mmc-core-set-initial-signal-voltage-on-power-off.patch
new file mode 100644
index 000000000..223059127
--- /dev/null
+++ b/target/linux/rockchip/patches-5.4/300-mmc-core-set-initial-signal-voltage-on-power-off.patch
@@ -0,0 +1,32 @@
+From bee5e8e3f9c5c7437711d7be8b36e1091b371847 Mon Sep 17 00:00:00 2001
+From: Jonas Karlman <jonas@kwiboo.se>
+Date: Tue, 24 Jul 2018 15:50:06 +0200
+Subject: [PATCH] mmc: core: set initial signal voltage on power off
+
+Signed-off-by: jensen <jensenhuang@friendlyarm.com>
+---
+ drivers/mmc/core/core.c | 8 ++++++++
+ 1 file changed, 8 insertions(+)
+
+diff --git a/drivers/mmc/core/core.c b/drivers/mmc/core/core.c
+index 26644b7ec13e..94534bc40bae 100644
+--- a/drivers/mmc/core/core.c
++++ b/drivers/mmc/core/core.c
+@@ -1366,6 +1366,14 @@ void mmc_power_off(struct mmc_host *host)
+ 	if (host->ios.power_mode == MMC_POWER_OFF)
+ 		return;
+ 
++	mmc_set_initial_signal_voltage(host);
++
++	/*
++	 * This delay should be sufficient to allow the power supply
++	 * to reach the minimum voltage.
++	 */
++	mmc_delay(10);
++
+ 	mmc_pwrseq_power_off(host);
+ 
+ 	host->ios.clock = 0;
+-- 
+2.25.1
+
diff --git a/target/linux/rockchip/patches-5.4/301-mmc-add-new-binding-to-specify-host-min-index.patch b/target/linux/rockchip/patches-5.4/301-mmc-add-new-binding-to-specify-host-min-index.patch
new file mode 100644
index 000000000..38289496c
--- /dev/null
+++ b/target/linux/rockchip/patches-5.4/301-mmc-add-new-binding-to-specify-host-min-index.patch
@@ -0,0 +1,38 @@
+From 659b50d846a18a4999026e01fd10c2129d7a56db Mon Sep 17 00:00:00 2001
+From: jensen <jensenhuang@friendlyarm.com>
+Date: Mon, 24 Aug 2020 20:07:11 +0800
+Subject: [PATCH] mmc: add new binding to specify host min index
+
+Signed-off-by: jensen <jensenhuang@friendlyarm.com>
+---
+ drivers/mmc/core/host.c | 7 ++++++-
+ 1 file changed, 6 insertions(+), 1 deletion(-)
+
+diff --git a/drivers/mmc/core/host.c b/drivers/mmc/core/host.c
+index b3484def0a8b..9cb982bc2349 100644
+--- a/drivers/mmc/core/host.c
++++ b/drivers/mmc/core/host.c
+@@ -397,6 +397,7 @@ struct mmc_host *mmc_alloc_host(int extra, struct device *dev)
+ {
+ 	int err;
+ 	struct mmc_host *host;
++	u32 index_min = 0;
+ 
+ 	host = kzalloc(sizeof(struct mmc_host) + extra, GFP_KERNEL);
+ 	if (!host)
+@@ -405,7 +406,11 @@ struct mmc_host *mmc_alloc_host(int extra, struct device *dev)
+ 	/* scanning will be enabled when we're ready */
+ 	host->rescan_disable = 1;
+ 
+-	err = ida_simple_get(&mmc_host_ida, 0, 0, GFP_KERNEL);
++	device_property_read_u32(dev, "host-index-min", &index_min);
++	if (index_min > 0)
++		dev_dbg(dev, "alloc index from %d\n", index_min);
++
++	err = ida_simple_get(&mmc_host_ida, index_min, 0, GFP_KERNEL);
+ 	if (err < 0) {
+ 		kfree(host);
+ 		return NULL;
+-- 
+2.25.1
+
diff --git a/target/linux/rockchip/patches-5.4/302-arm64-dts-rockchip-Add-Nanopi-R4S-board-support.patch b/target/linux/rockchip/patches-5.4/302-arm64-dts-rockchip-Add-Nanopi-R4S-board-support.patch
new file mode 100644
index 000000000..73ba08827
--- /dev/null
+++ b/target/linux/rockchip/patches-5.4/302-arm64-dts-rockchip-Add-Nanopi-R4S-board-support.patch
@@ -0,0 +1,239 @@
+From 9dd819240ddbff440e758e5737cc425d06cf61d5 Mon Sep 17 00:00:00 2001
+From: jensen <jensenhuang@friendlyarm.com>
+Date: Sun, 20 Dec 2020 18:16:44 +0800
+Subject: [PATCH] arm64: dts: rockchip: Add Nanopi R4S board support
+
+Specification
+- Rockchip RK3399
+- LPDDR4 4GiB, 1GB DDR3-1866
+- SD card slot
+- RTL8211E 1Gbps
+- RTL8111H 1Gbps
+- USB 3.0 x2
+- USB Type C power
+
+Change-Id: I4de7f80c29cd9028ba99c56fbe9ac7f68cd356af
+Signed-off-by: jensen <jensenhuang@friendlyarm.com>
+---
+ arch/arm64/boot/dts/rockchip/Makefile         |   1 +
+ .../boot/dts/rockchip/rk3399-nanopi-r4s.dts   | 153 ++++++++++++++++++
+ .../boot/dts/rockchip/rk3399-nanopi4.dtsi     |  10 +-
+ 3 files changed, 161 insertions(+), 3 deletions(-)
+ create mode 100644 arch/arm64/boot/dts/rockchip/rk3399-nanopi-r4s.dts
+
+diff --git a/arch/arm64/boot/dts/rockchip/Makefile b/arch/arm64/boot/dts/rockchip/Makefile
+index 1f18a9392d15..5489c9db69b9 100644
+--- a/arch/arm64/boot/dts/rockchip/Makefile
++++ b/arch/arm64/boot/dts/rockchip/Makefile
+@@ -24,6 +24,7 @@ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-leez-p710.dtb
+ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-nanopc-t4.dtb
+ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-nanopi-m4.dtb
+ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-nanopi-neo4.dtb
++dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-nanopi-r4s.dtb
+ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-orangepi.dtb
+ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-puma-haikou.dtb
+ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-roc-pc.dtb
+diff --git a/arch/arm64/boot/dts/rockchip/rk3399-nanopi-r4s.dts b/arch/arm64/boot/dts/rockchip/rk3399-nanopi-r4s.dts
+new file mode 100644
+index 000000000000..4d16bdb885b2
+--- /dev/null
++++ b/arch/arm64/boot/dts/rockchip/rk3399-nanopi-r4s.dts
+@@ -0,0 +1,153 @@
++// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
++/*
++ * FriendlyElec NanoPi R4S board device tree source
++ *
++ * Copyright (c) 2020 FriendlyElec Computer Tech. Co., Ltd.
++ * (http://www.friendlyarm.com)
++ *
++ * Copyright (c) 2018 Collabora Ltd.
++ */
++
++/dts-v1/;
++#include "rk3399-nanopi4.dtsi"
++
++/ {
++	model = "FriendlyElec NanoPi R4S";
++	compatible = "friendlyarm,nanopi-r4s", "rockchip,rk3399";
++
++	aliases {
++		ethernet1 = &r8169;
++	};
++
++	vdd_5v: vdd-5v {
++		compatible = "regulator-fixed";
++		regulator-name = "vdd_5v";
++		regulator-always-on;
++		regulator-boot-on;
++	};
++
++	fan: pwm-fan {
++		compatible = "pwm-fan";
++		/* FIXME: adjust leveles for the connected fan */
++		cooling-levels = <0 12 18 255>;
++		#cooling-cells = <2>;
++		fan-supply = <&vdd_5v>;
++		pwms = <&pwm1 0 50000 0>;
++	};
++};
++
++&cpu_thermal {
++	trips {
++		cpu_warm: cpu_warm {
++			temperature = <55000>;
++			hysteresis = <2000>;
++			type = "active";
++		};
++
++		cpu_hot: cpu_hot {
++			temperature = <65000>;
++			hysteresis = <2000>;
++			type = "active";
++		};
++	};
++
++	cooling-maps {
++		map2 {
++			trip = <&cpu_warm>;
++			cooling-device = <&fan THERMAL_NO_LIMIT 1>;
++		};
++
++		map3 {
++			trip = <&cpu_hot>;
++			cooling-device = <&fan 2 THERMAL_NO_LIMIT>;
++		};
++	};
++};
++
++&emmc_phy {
++	status = "disabled";
++};
++
++&fusb0 {
++	status = "disabled";
++};
++
++&keys {
++	button@1 {
++		debounce-interval = <50>;
++		gpios = <&gpio1 RK_PC6 GPIO_ACTIVE_LOW>;
++		label = "K1";
++		linux,code = <BTN_1>;
++		wakeup-source;
++	};
++};
++
++&key_pins {
++	rockchip,pins =
++		<0 RK_PA5 RK_FUNC_GPIO &pcfg_pull_up>,
++		<1 RK_PC6 RK_FUNC_GPIO &pcfg_pull_up>;
++};
++
++&leds {
++	lan_led: led-1 {
++		gpios = <&gpio1 RK_PA1 GPIO_ACTIVE_HIGH>;
++		label = "lan_led";
++	};
++
++	wan_led: led-2 {
++		gpios = <&gpio1 RK_PA0 GPIO_ACTIVE_HIGH>;
++		label = "wan_led";
++	};
++};
++
++&leds_gpio {
++	rockchip,pins =
++		<0 RK_PB5 RK_FUNC_GPIO &pcfg_pull_none>,
++		<1 RK_PA0 RK_FUNC_GPIO &pcfg_pull_none>,
++		<1 RK_PA1 RK_FUNC_GPIO &pcfg_pull_none>;
++};
++
++&pcie0 {
++	max-link-speed = <1>;
++	num-lanes = <1>;
++	vpcie3v3-supply = <&vcc3v3_sys>;
++
++	pcie@0 {
++		reg = <0x00000000 0 0 0 0>;
++		#address-cells = <3>;
++		#size-cells = <2>;
++
++		r8169: pcie@0,0 {
++			reg = <0x000000 0 0 0 0>;
++			local-mac-address = [ 00 00 00 00 00 00 ];
++		};
++	};
++};
++
++&sdhci {
++	status = "disabled";
++};
++
++&sdio0 {
++	status = "disabled";
++};
++
++&sdmmc {
++	host-index-min = <1>;
++};
++
++&u2phy0_host {
++	phy-supply = <&vdd_5v>;
++};
++
++&u2phy1_host {
++	status = "disabled";
++};
++
++&usbdrd_dwc3_0 {
++	dr_mode = "host";
++};
++
++&vcc3v3_sys {
++	vin-supply = <&vcc5v0_sys>;
++};
+diff --git a/arch/arm64/boot/dts/rockchip/rk3399-nanopi4.dtsi b/arch/arm64/boot/dts/rockchip/rk3399-nanopi4.dtsi
+index dd16c80d923e..092c10a27c20 100644
+--- a/arch/arm64/boot/dts/rockchip/rk3399-nanopi4.dtsi
++++ b/arch/arm64/boot/dts/rockchip/rk3399-nanopi4.dtsi
+@@ -78,11 +78,11 @@
+ 		regulator-name = "vbus_typec";
+ 	};
+ 
+-	gpio-keys {
++	keys: gpio-keys {
+ 		compatible = "gpio-keys";
+ 		autorepeat;
+ 		pinctrl-names = "default";
+-		pinctrl-0 = <&power_key>;
++		pinctrl-0 = <&key_pins>;
+ 
+ 		power {
+ 			debounce-interval = <100>;
+@@ -519,7 +519,7 @@
+ 	};
+ 
+ 	rockchip-key {
+-		power_key: power-key {
++		key_pins: key-pins {
+ 			rockchip,pins = <0 RK_PA5 RK_FUNC_GPIO &pcfg_pull_up>;
+ 		};
+ 	};
+@@ -685,10 +685,14 @@
+ };
+ 
+ &usbdrd_dwc3_0 {
++	snps,xhci-slow-suspend-quirk;
++	snps,xhci-trb-ent-quirk;
+ 	status = "okay";
+ };
+ 
+ &usbdrd_dwc3_1 {
++	snps,xhci-slow-suspend-quirk;
++	snps,xhci-trb-ent-quirk;
+ 	dr_mode = "host";
+ 	status = "okay";
+ };
+-- 
+2.25.1
+
-- 
2.25.1

